ecosystem/indexer-grpc/indexer-grpc-table-info/src/table_info_parser.rs

```diff
@@ -0,0 +1,323 @@
+                            batch.num_transactions_to_fetch, retries, batch.start_version, err
+                        );
+                    } else {
+                        error!(
+                            starting_version = batch.start_version,
+                            num_transactions = batch.num_transactions_to_fetch,
+                            error = format!("{:?}", err),
+                            "[Table Info] Could not fetch transactions: will retry",
+                        );
+                    }
+                    tokio::time::sleep(Duration::from_millis(300)).await;
+                },
+            }
+        }
+    }
+
+    fn parse_table_info(
+        context: Arc<Context>,
+        raw_txns: Vec<TransactionOnChainData>,
+        db_writer: Arc<dyn DbWriter>,
+        end_early_if_pending_on_empty: bool,
+    ) -> Result<(), Error> {
+        if raw_txns.is_empty() {
+            return Ok(());
+        }
+
+        let start_millis = chrono::Utc::now().naive_utc();
+        let first_version = raw_txns.first().map(|txn| txn.version).unwrap();
+
+        let write_sets: Vec<WriteSet> = raw_txns.iter().map(|txn| txn.changes.clone()).collect();
+        let write_sets_slice: Vec<&WriteSet> = write_sets.iter().collect();
+        db_writer
+            .clone()
+            .index_table_info(
+                context.db.clone(),
+                first_version,
+                &write_sets_slice,
+                end_early_if_pending_on_empty,
+            )
+            .expect(
+                "[Table Info] Failed to process write sets and index to the table info rocksdb",
+            );
+        let fetch_millis = (chrono::Utc::now().naive_utc() - start_millis).num_milliseconds();
+
+        info!(
+            table_info_first_version = first_version,
+            write_sets_bcs_size = bcs::serialized_size(&write_sets).unwrap(),
+            table_info_parsing_millis_per_batch = fetch_millis,
+            num_transactions = raw_txns.len(),
+            "[Table Info] Table info parsed successfully"
+        );
+
+        Ok(())
+    }
+
+    /// Will keep looping and checking the latest ledger info to see if there are new transactions
+    /// If there are, it will update the ledger version version
+    async fn get_highest_known_version(&mut self) -> Result<u64, Error> {
+        let mut info = self.context.get_latest_ledger_info_wrapped();
+        let mut ledger_version = info.unwrap().ledger_version.0;
+        let mut empty_loops = 0;
+        while ledger_version == 0 || self.current_version > ledger_version {
+            if empty_loops > 0 {
+                tokio::time::sleep(Duration::from_millis(LEDGER_VERSION_RETRY_TIME_MILLIS)).await;
+            }
+            empty_loops += 1;
+            if let Err(err) =
+                {
+                    info = self.context.get_latest_ledger_info_wrapped();
```

Comment:
Can we add a todo here to consolidate with the other function? If we replace `self` with `context` we can actually share the function.